File: ruby/problem15.rb
$pathCache = {}

# x = distance from right
# y = distance from bottom
def getNumPaths(x, y)
  if x == 0 || y == 0
    return 1
  elsif $pathCache[[x,y]]
    $pathCache[[x,y]]
  else
    $pathCache[[x,y]] = getNumPaths(x - 1, y) + getNumPaths(x, y - 1)
    $pathCache[[y,x]] = $pathCache[[x,y]] # symmetrical points have same answer
  end
end

puts getNumPaths(20, 20)

File: javascript/problem15.js
var cache = {
  '0,0': 1,
};

function latticePaths(x, y) {
  if ( cache[ [ x, y ] ] ) { // handles (20,20)
    return cache[ [ x, y ] ];
  } else if ( x == 0 ) {
    cache[ [ x, y ] ] = latticePaths( x, y - 1 );
  } else if ( y == 0 ) {
    cache[ [ x, y ] ] = latticePaths( x - 1, y );
  } else {
    var down = latticePaths( x, y - 1 );
    var right = latticePaths( x - 1, y );
    cache[ [ x, y ] ] = down + right;
    cache[ [ y, x ] ] = down + right; // symmetric points have same answer
  }

  return cache[ [ x, y ] ];
}

console.log( latticePaths( 20, 20 ) );
